api/envoy/extensions/transport_sockets/tls/v3/tls.proto

```diff
@@ -127,8 +145,15 @@ message CommonTlsContext {
   repeated TlsCertificate tls_certificates = 2;
 
   // Configs for fetching TLS certificates via SDS API.
-  repeated SdsSecretConfig tls_certificate_sds_secret_configs = 6
-      [(validate.rules).repeated = {max_items: 1}];
+  repeated SdsSecretConfig tls_certificate_sds_secret_configs = 6 [
+    (validate.rules).repeated = {max_items: 1},
+    (udpa.annotations.field_migrate).oneof_promotion = "dynamic_tls_certificate"
+  ];
+
+  // Config for fetching TLS certificates via CA client plugin.
+  // [#not-implemented-hide:]
+  CAClientPluginConfig tls_certificate_ca_client_plugin = 9
```

Comment:
Is the expectation that `validation_context_ca_client_plugin` and `tls_certificate_ca_client_plugin` have the same config? If so, should this be written explicitly. I.e. are we going to instantiate the plugin twice, or just once via a dedupe?

Reply:
> Is the expectation that `validation_context_ca_client_plugin` and `tls_certificate_ca_client_plugin` have the same config?

In the simple case of a single CA, yes, the `validation_context_ca_client_plugin` and `tls_certificate_ca_client_plugin` will have the same config value. But in more complex cases (e.g. a federated service mesh where a client connects to a server using a different CA) the 2 could be different.

>If so, should this be written explicitly. I.e. are we going to instantiate the plugin twice, or just once via a dedupe?

That's really an implementation choice, is it not? It makes sense to dedup it for various reasons and the gRPC implementation will most likely do it. What do we mention here about that?

Reply:
At the end of the day the docs exist for operators to know how to configure their system. So, as long as it's clear to the operator what they need to do, e.g. set consistent options for the extension, then all is good.

Reply:
Something I should have raised earlier: in case of SDS even if the rest of the SDS config is the same between `tls_certificate_sds_secret_configs` and `validation_context_sds_secret_configs` there is still a [name](https://github.com/envoyproxy/envoy/blob/master/api/envoy/extensions/transport_sockets/tls/v3/secret.proto#L33) that is used by the control plane to differentiate between local cert and a (root) CA cert request. In this new proposal we don't have anything similar. So if 
 `tls_certificate_ca_client_plugin` and `validation_context_ca_client_plugin` are the same the CAClientPlugin will need an additional flag stored in its cache to use these configs appropriately. Does it then make sense to retain the `CAClientPluginConfig` struct and the name there can be used similar to [name](https://github.com/envoyproxy/envoy/blob/master/api/envoy/extensions/transport_sockets/tls/v3/secret.proto#L33) in `SdsSecretConfig`?

Reply:
Can we just use the opaque name in `TypedExtensionConfig` for this?

Reply:
I think we're talking about two separate names here:

1. The name of the certificate to request.  (In SDS, this was essentially the SDS resource name to request.)
2. The name of the certificate provider plugin to use.

For (2), I think it should be encoded in the `name` field in the `TypedExtensionConfig` message.

For (1), I see two options:

A. We can keep the `CAClientPluginConfig` message, and have its `name` field indicate (1).

B. We can not encode this name explicitly in the xDS API but instead require individual CA client plugins to encode this in their own configs (i.e., in the `Any` field inside of `TypedExtensionConfig`).

I think (A) makes sense if we expect all plugin implementations to need this kind of name.  If we think there are some plugins that won't need this kind of name, then we should go with (B).

Reply:
Yeah, I prefer (B) unless almost all extensions will need this name. Keep in mind that the name format may be different in each, may have more or less structure that is reflected at a proto level, etc.

Reply:
I prefer (A) for 2 reasons:

-  that way `CAClientPluginConfig.typed_config` can be treated as static config for the plugin (e.g. target_uri, cred to use etc) and  `CAClientPluginConfig.name` is the dynamic part (identifying cert resource similar to `SdsSecretConfig.name`
- almost all plugins would use a name for cert id

So I am going to leave it for now.

Reply:
Yeah, I think most plugins will want some kind of naming for certificates, but will it always be an opaque string? Might it be a more structured proto-ificaiton of an address or URI? This is the rationale that leads me towards (B). I think if we want to go down the route of (A), it would be helpful for you to walk through a reasonable variety of different plugins in the design doc and basically show they all want string.

This is the only outstanding thread from my perspective in this PR. The current comment on local vs. root is a bit confusing as well.

Reply:
I removed the local/root part from the comment and clarified with an example. Regarding the naming of certificates, this is what I am trying to say: the certificate-name is really a "handle" between the xDS server and the xDS client as a convenient way to reference certificate instances or certificate types and not really dependent on the plugin (or the CA). I will illustrate with a couple of examples:
- as a certificate-type the xDS server can use the name "ROOTCA" to specify a root certificate (for validation contexts) or "TLS-LOCAL" to specify a tls-certificate to be issued to the local client (or server).
- as a certificate-instance - and also extending the above scheme - the xDS server can use the name "TLS-LOCAL1" to specify a tls-certificate to be issued to the local client talking to cluster1 and another name "TLS-LOCAL2" to specify another tls-certificate to be issued to the local client talking to cluster2. And "TLS-LOCAL3" for all the servers/listeners in the application. This way the xDS server can use the certificate-name to specify which exact certificates are to be shared or not shared among a bunch of clients or servers (within a gRPC application for example).

As mentioned earlier this is used somewhat like a resource name that uniquely identifies a certificate resource (within an application). I'll update the design doc if others think this makes sense.

Reply:
In your example, does the client need to be aware of the root vs. local naming convention?